specs/chugsplash-manager.md

```diff
@@ -0,0 +1,165 @@
+      bundleRoot,
+      bundleSize,
+      configUri
+    )
+  )
+}
+```
+
+## Getting the selected executor
+
+```typescript
+const getSelectedExecutor = (
+  bundleId: bytes32
+): address => {
+  return bundles[bundleId].selectedExecutor
+}
+```
+
+## Getting the proxy by name
+
+```typescript
+const getProxyByName = (
+  name: string
+): address payable => {
+  return (
+    payable(
+      Create2.compute(address(this), keccak256(bytes(name)), type(Proxy).creationCode)
+    )
+  )
+}
+```
+
+## Proposing a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be proposable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST be in the `EMPTY` state before they can be proposed.
+* Proposing a bundle puts the bundle in the `PROPOSED` state.
+* The ChugSplashManager MUST emit the event ChugSplashBundleProposed with the ID of the bundle, the bundle root, the bundle size, and the config URI as parameters. The ChugSplashManager MUST announce this event to the registry.
+
+## Approving a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be approvable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST have be in the `PROPOSED` state before they can be approved.
+* There MUST NOT be any active bundle.
+* Approving a bundle MUST put the bundle in the `APPROVED` state.
+* Approving a bundle MUST make the approved bundle the active bundle.
+* The balance of the ChugSplashManager minus the current totalDebt MUST be greater than BUNDLE_EXECUTION_BOND_AMOUNT ETH in its balance.
+* The ChugSplashManager MUST emit the event ChugSplashBundleApproved with the ID of the bundle as the only parameter. The ChugSplashManager MUST announce this event to the registry.
+
+## Executing a ChugSplash action
+* There MUST be an `activeBundleId` to execute a ChugSplash action.
+* There MUST be enough reserved ETH in the ChugSplashManager to pay the executor for this action.
+* The `_actionIndex` MUST NOT already have been executed.
+* A ChugSplash action MUST be executed by the `selectedExecutor`.
+* The `_action`, `activeBundleId`, `_actionIndex`, and `_proof` MUST produce a valid Merkle proof.
+* If the proxy corresponding to the `target` has not already been deployed:
+  * The proxy MUST be deployed using Create2.
+  * The proxy MUST have the same address as the the address returned by a call to `getProxyByName`.
+* A `SET_CODE` or `SET_ACTION` MUST be executed on the proxy depending on the `ActionType`.
+* Executing an action MUST increment the `actionsExecuted` by one.
+* Executing an action MUST set the current `_actionIndex` to `true`.
+* The ChugSplashManager MUST emit the event ChugSplashActionExecuted with the active bundle ID, the executor's address, and the action index. The ChugSplashManager MUST announce this event to the registry.
+* If all actions have been executed for the bundle:
+  * The bundle status MUST be set to `COMPLETED`.
+  * The ChugSplashManager MUST increase the `debt` owed to the current executor by the `executorBondAmount`.
```

Comment:
I dont see the debt variables in the existing contract, but I agree this is a critical feature, and the flow of debt and repayment should be carefully mapped out.

Reply:
Agreed, we're going to finalize this spec before updating the contract